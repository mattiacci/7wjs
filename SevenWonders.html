<script src="config.js"></script>
<script>
window.ASSET_URL_PREFIX = window.ASSET_URL_PREFIX || '';
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/react/0.14.0/react.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/react/0.14.0/react-dom.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.6.15/browser.js"></script>
<script src="https://cdn.firebase.com/js/client/2.3.2/firebase.js"></script>
<script src="https://cdn.firebase.com/libs/reactfire/0.5.1/reactfire.min.js"></script>
<script src="card.jsx" type="text/babel"></script>
<script src="hand.jsx" type="text/babel"></script>
<script src="wonderboard.jsx" type="text/babel"></script>
<script src="playerfield.jsx" type="text/babel"></script>
<script src="lobby.jsx" type="text/babel"></script>
<script src="game.jsx" type="text/babel"></script>
<script src="app.jsx" type="text/babel"></script>
<script src="SevenWonders.js"></script>
<style>
  #legacy {
    display: flex;
    flex-direction: row-reverse;
    overflow-x: scroll;
    white-space: nowrap;
  }
  .card {
    border-radius: 12px; /* used for highlights or when images aren't available */
    box-sizing: border-box;
    display: inline-block;
    font-size: 12px;
    height: 158px;
    margin: 5px 2px;
    padding: 6px;
    position: relative;
    transition: .2s all;
    vertical-align: text-top;
    white-space: normal;
    width: 102px;
  }
  .card:hover {
    transform: scale(1.5);
    z-index: 2;
  }
  .card .image {
    background: center no-repeat;
    background-size: contain;
    border-radius: 12px; /* used to crop image */
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
  }
  .card.selected .image {
    border-radius: 12px;
    box-shadow: 0 0 0px 3px red;
    z-index: 1;
  }
  .card.unplayable:not(:hover) {
    -webkit-filter: grayscale(50%) opacity(60%);
    filter: grayscale(50%) opacity(60%);
  }
  .card.free {
    box-shadow: 0 0 0px 3px gold;
  }
  .board {
    background-size: contain;
    background-position: top;
    background-repeat: no-repeat;
    width: 510px;
    height: 224px;
    position: absolute;
  }
  .back {
    position: absolute;
    border-radius: 12px;
    width: 128px;
    height: 197px;
    top: 169px;
    z-index: 0;
    background-size: contain;
    background-repeat: no-repeat;
  }
  .wonder {
    margin-left: 15px;
    width: 510px;
    height: 390px;
    position: relative;
  }
</style>
<div id="app"></div>
<div id="score"></div>
<div id="legacy"></div>
<script>
  (function() {
    var params = (function() {
      var result = {};
      if (window.location.search) {
          // split up the query string and store in an associative array
          var params = window.location.search.slice(1).split('&');
          for (var i = 0; i < params.length; i++) {
              var tmp = params[i].split('=');
              result[tmp[0]] = unescape(tmp[1]);
          }
      }
      return result;
    }());

    function go() {
      // TODO: Add browserify or similar.
      if (!window.App) {
        return false;
      }

      window.gameRoom = window.gameRoom ||
          new GameRoom(document.getElementById('app'), document.getElementById('legacy'));

      if (params['game']) {
        // TODO: Handle bookmarking properly.
        var gameButton = document.querySelector('button[data-game="' + params['game'] + '"]');
        if (gameButton == null) {
          return false;
        } else {
          gameButton.click();
        }
      }

      return true;
    }

    function retryUntilTruthy(func) {
      if (!func()) {
        window.setTimeout(function() {
          return retryUntilTruthy(func);
        }, 100);
      }
    }

    retryUntilTruthy(go);
  })();
</script>
