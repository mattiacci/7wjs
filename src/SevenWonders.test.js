import SevenWonders from './SevenWonders.js';

const WONDERS = [{"name":"Rhodos","side":"A"},{"name":"Ephesos","side":"A"},{"name":"Olympia","side":"A"}];
const CARDS = [[[{"age":1,"minPlayers":3,"name":"Clay Pool"},{"age":1,"minPlayers":2,"name":"Stockade"},{"age":1,"minPlayers":3,"name":"Guard Tower"},{"age":1,"minPlayers":2,"name":"Timber Yard"},{"age":1,"minPlayers":2,"name":"Press"},{"age":1,"minPlayers":2,"name":"Barracks"},{"age":1,"minPlayers":2,"name":"Workshop"}],[{"age":1,"minPlayers":2,"name":"Apothecary"},{"age":1,"minPlayers":2,"name":"West Trading Post"},{"age":1,"minPlayers":2,"name":"Marketplace"},{"age":1,"minPlayers":2,"name":"Theater"},{"age":1,"minPlayers":2,"name":"Loom"},{"age":1,"minPlayers":2,"name":"Altar"},{"age":1,"minPlayers":2,"name":"Glassworks"}],[{"age":1,"minPlayers":2,"name":"Scriptorium"},{"age":1,"minPlayers":2,"name":"Clay Pit"},{"age":1,"minPlayers":3,"name":"East Trading Post"},{"age":1,"minPlayers":3,"name":"Ore Vein"},{"age":1,"minPlayers":3,"name":"Baths"},{"age":1,"minPlayers":3,"name":"Lumber Yard"},{"age":1,"minPlayers":3,"name":"Stone Pit"}]],[[{"age":2,"minPlayers":3,"name":"Press"},{"age":2,"minPlayers":3,"name":"Aquaduct"},{"age":2,"minPlayers":2,"name":"Foundry"},{"age":2,"minPlayers":3,"name":"Walls"},{"age":2,"minPlayers":2,"name":"Stables"},{"age":2,"minPlayers":2,"name":"Archery Range"},{"age":2,"minPlayers":3,"name":"School"}],[{"age":2,"minPlayers":2,"name":"Forum"},{"age":2,"minPlayers":3,"name":"Vineyard"},{"age":2,"minPlayers":2,"name":"Brickyard"},{"age":2,"minPlayers":2,"name":"Temple"},{"age":2,"minPlayers":2,"name":"Sawmill"},{"age":2,"minPlayers":2,"name":"Dispensary"},{"age":2,"minPlayers":3,"name":"Loom"}],[{"age":2,"minPlayers":2,"name":"Statue"},{"age":2,"minPlayers":2,"name":"Caravansery"},{"age":2,"minPlayers":2,"name":"Library"},{"age":2,"minPlayers":2,"name":"Laboratory"},{"age":2,"minPlayers":2,"name":"Quarry"},{"age":2,"minPlayers":2,"name":"Courthouse"},{"age":2,"minPlayers":3,"name":"Glassworks"}]],[[{"age":3,"minPlayers":2,"name":"Arena"},{"age":3,"minPlayers":2,"name":"Pantheon"},{"age":3,"minPlayers":2,"name":"Gardens"},{"age":3,"minPlayers":3,"name":"Fortifications"},{"age":3,"minPlayers":3,"name":"Academy"},{"age":3,"minPlayers":2,"name":"Lighthouse"},{"age":3,"minPlayers":0,"name":"Spies Guild"}],[{"age":3,"minPlayers":0,"name":"Craftsmens Guild"},{"age":3,"minPlayers":0,"name":"Strategists Guild"},{"age":3,"minPlayers":2,"name":"Observatory"},{"age":3,"minPlayers":3,"name":"Palace"},{"age":3,"minPlayers":3,"name":"Arsenal"},{"age":3,"minPlayers":0,"name":"Traders Guild"},{"age":3,"minPlayers":3,"name":"Town Hall"}],[{"age":3,"minPlayers":0,"name":"Builders Guild"},{"age":3,"minPlayers":2,"name":"Senate"},{"age":3,"minPlayers":2,"name":"Siege Workshop"},{"age":3,"minPlayers":3,"name":"Study"},{"age":3,"minPlayers":2,"name":"Lodge"},{"age":3,"minPlayers":2,"name":"University"},{"age":3,"minPlayers":2,"name":"Haven"}]]];

it('starts game and passes state and actions only to local player', () => {
  const playerInterfaces = [];
  const stateUpdateHandler = jest.fn();
  const remotePlayerStateUpdateHandler = jest.fn();
  const turnsRef = [];
  playerInterfaces.push(new SevenWonders.PlayerInterface(
      stateUpdateHandler, turnsRef, 0, '0', true));
  for (let i = 1; i < WONDERS.length; i++) {
    playerInterfaces.push(new SevenWonders.PlayerInterface(
        remotePlayerStateUpdateHandler, turnsRef, i, String(i), false));
  }
  const endGameHandler = jest.fn();
  const game = new SevenWonders(playerInterfaces, WONDERS, CARDS, false, endGameHandler);
  expect(endGameHandler).not.toHaveBeenCalled();
  expect(remotePlayerStateUpdateHandler).not.toHaveBeenCalled();
  expect(turnsRef).toHaveLength(0);
  expect(stateUpdateHandler).toHaveBeenCalledTimes(1);
  const state = stateUpdateHandler.mock.calls[0][0];
  expect(state).toMatchSnapshot('initial');
  const actions = stateUpdateHandler.mock.calls[0][1];
  expect(actions).toHaveProperty('build');
  expect(actions).toHaveProperty('buildWonder');
  expect(actions).toHaveProperty('discard');
});
