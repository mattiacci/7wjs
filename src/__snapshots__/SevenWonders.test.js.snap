// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial 1`] = `
Object {
  "game": Object {
    "age": 0,
  },
  "players": Array [
    Object {
      "battleTokens": Array [],
      "built": Array [],
      "cardsInHand": Array [
        Object {
          "age": 1,
          "cost": Array [],
          "free": true,
          "minPlayers": 3,
          "name": "Clay Pool",
          "tooltip": "Produces clay.",
          "type": 0,
          "unplayable": false,
        },
        Object {
          "age": 1,
          "cost": Array [
            Array [
              2,
            ],
          ],
          "free": false,
          "minPlayers": 2,
          "name": "Stockade",
          "tooltip": "Provides 1 unit of military strength.",
          "type": 4,
          "unplayable": false,
        },
        Object {
          "age": 1,
          "cost": Array [
            Array [
              0,
            ],
          ],
          "free": false,
          "minPlayers": 3,
          "name": "Guard Tower",
          "tooltip": "Provides 1 unit of military strength.",
          "type": 4,
          "unplayable": true,
        },
        Object {
          "age": 1,
          "cost": Array [
            1,
          ],
          "free": false,
          "minPlayers": 2,
          "name": "Timber Yard",
          "tooltip": "Produces stone or wood.",
          "type": 0,
          "unplayable": false,
        },
        Object {
          "age": 1,
          "cost": Array [],
          "free": true,
          "minPlayers": 2,
          "name": "Press",
          "tooltip": "Produces paper.",
          "type": 1,
          "unplayable": false,
        },
        Object {
          "age": 1,
          "cost": Array [
            Array [
              3,
            ],
          ],
          "free": true,
          "minPlayers": 2,
          "name": "Barracks",
          "tooltip": "Provides 1 unit of military strength.",
          "type": 4,
          "unplayable": false,
        },
        Object {
          "age": 1,
          "cost": Array [
            Array [
              4,
            ],
          ],
          "free": false,
          "minPlayers": 2,
          "name": "Workshop",
          "tooltip": "Enhances engineering.",
          "type": 6,
          "unplayable": true,
        },
      ],
      "gold": 3,
      "id": 0,
      "name": "0",
      "playable": true,
      "resources": Object {
        "east": Object {
          "multi": Array [],
          "single": Object {
            "0": 0,
            "1": 0,
            "2": 1,
            "3": 0,
            "4": 0,
            "5": 0,
            "6": 0,
          },
        },
        "multi": Array [],
        "single": Object {
          "0": 0,
          "1": 0,
          "2": 0,
          "3": 1,
          "4": 0,
          "5": 0,
          "6": 0,
        },
        "west": Object {
          "multi": Array [],
          "single": Object {
            "0": 0,
            "1": 0,
            "2": 0,
            "3": 0,
            "4": 0,
            "5": 0,
            "6": 1,
          },
        },
      },
      "score": 1,
      "waiting": false,
      "wonder": Object {
        "built": Array [],
        "isLast": false,
        "name": "Rhodos",
        "side": "A",
        "stageCount": 5,
      },
    },
    Object {
      "battleTokens": Array [],
      "built": Array [],
      "gold": 3,
      "id": 1,
      "name": "1",
      "playable": false,
      "resources": Object {
        "east": Object {
          "multi": Array [],
          "single": Object {
            "0": 0,
            "1": 0,
            "2": 0,
            "3": 1,
            "4": 0,
            "5": 0,
            "6": 0,
          },
        },
        "multi": Array [],
        "single": Object {
          "0": 0,
          "1": 0,
          "2": 0,
          "3": 0,
          "4": 0,
          "5": 0,
          "6": 1,
        },
        "west": Object {
          "multi": Array [],
          "single": Object {
            "0": 0,
            "1": 0,
            "2": 1,
            "3": 0,
            "4": 0,
            "5": 0,
            "6": 0,
          },
        },
      },
      "score": 1,
      "waiting": false,
      "wonder": Object {
        "built": Array [],
        "isLast": false,
        "name": "Ephesos",
        "side": "A",
        "stageCount": 6,
      },
    },
    Object {
      "battleTokens": Array [],
      "built": Array [],
      "gold": 3,
      "id": 2,
      "name": "2",
      "playable": false,
      "resources": Object {
        "east": Object {
          "multi": Array [],
          "single": Object {
            "0": 0,
            "1": 0,
            "2": 0,
            "3": 0,
            "4": 0,
            "5": 0,
            "6": 1,
          },
        },
        "multi": Array [],
        "single": Object {
          "0": 0,
          "1": 0,
          "2": 1,
          "3": 0,
          "4": 0,
          "5": 0,
          "6": 0,
        },
        "west": Object {
          "multi": Array [],
          "single": Object {
            "0": 0,
            "1": 0,
            "2": 0,
            "3": 1,
            "4": 0,
            "5": 0,
            "6": 0,
          },
        },
      },
      "score": 1,
      "waiting": false,
      "wonder": Object {
        "built": Array [],
        "isLast": false,
        "name": "Olympia",
        "side": "A",
        "stageCount": 6,
      },
    },
  ],
}
`;
